File: jApp/src/main/java/j/s/yarlykov/ui/fragmentbased/TemperatureSensorFragment.java
package j.s.yarlykov.ui.fragmentbased;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import j.s.yarlykov.R;

public class TemperatureSensorFragment extends Fragment implements SensorEventListener {
    private SensorManager sensorManager;
    private Sensor sensorTemperature = null;
    private Sensor sensorHumidity = null;
    private TextView tvT, tvH;

    public static TemperatureSensorFragment create() {
        TemperatureSensorFragment fragment = new TemperatureSensorFragment();
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        sensorTemperature = sensorManager.getDefaultSensor(Sensor.TYPE_AMBIENT_TEMPERATURE);
        sensorHumidity = sensorManager.getDefaultSensor(Sensor.TYPE_RELATIVE_HUMIDITY);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.current_temperature_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        tvT = view.findViewById(R.id.tvT);
        tvH = view.findViewById(R.id.tvH);

        String notSupported = getString(R.string.sensorNotSupported);

        if (sensorTemperature == null) {
            tvT.setText(notSupported);
        }

        if (sensorHumidity == null) {
            tvH.setText(notSupported);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        sensorManager.registerListener(this, sensorTemperature,
                SensorManager.SENSOR_DELAY_NORMAL);
        sensorManager.registerListener(this, sensorHumidity,
                SensorManager.SENSOR_DELAY_NORMAL);
    }

    @Override
    public void onPause() {
        super.onPause();
        sensorManager.unregisterListener(this);
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        switch (event.sensor.getType()) {
            case Sensor.TYPE_AMBIENT_TEMPERATURE:
                tvT.setText(String.valueOf(event.values[0]) + " \u2103");
                break;
            case Sensor.TYPE_RELATIVE_HUMIDITY:
                tvH.setText(String.valueOf(event.values[0]));

                break;
            default:

        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }
}

File: kApp/src/main/java/k/s/yarlykov/ui/fragmentbased/TemperatureSensorFragment.kt
package k.s.yarlykov.ui.fragmentbased

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import k.s.yarlykov.R
import kotlinx.android.synthetic.main.current_temperature_fragment.*

class TemperatureSensorFragment : Fragment(), SensorEventListener {
    lateinit var sensorManager: SensorManager
    var sensorTemperature: Sensor? = null
    var sensorHumidity: Sensor? = null

    companion object {
        fun create(): TemperatureSensorFragment {
            return TemperatureSensorFragment()
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        sensorManager = context!!.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sensorTemperature = sensorManager.getDefaultSensor(Sensor.TYPE_AMBIENT_TEMPERATURE)
        sensorHumidity = sensorManager.getDefaultSensor(Sensor.TYPE_RELATIVE_HUMIDITY)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.current_temperature_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val notSupported = getString(R.string.sensorNotSupported)

        if (sensorTemperature == null) {
            tvT.text = notSupported
        }
        if (sensorHumidity == null) {
            tvH.text = notSupported
        }
    }

    override fun onResume() {
        super.onResume()
        sensorManager.registerListener(this, sensorTemperature,
                SensorManager.SENSOR_DELAY_NORMAL)
        sensorManager.registerListener(this, sensorHumidity,
                SensorManager.SENSOR_DELAY_NORMAL)
    }

    override fun onPause() {
        super.onPause()
        sensorManager.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_AMBIENT_TEMPERATURE -> {
                val sT = "${event.values[0]} ℃"
                tvT.text = sT
            }
            Sensor.TYPE_RELATIVE_HUMIDITY -> {
                val sH = "${event.values[0]} %"
                tvH.text = sH
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}

}
